### PR TITLE
ARROW-8853: [Rust] [Integration Testing] Enable Flight tests

### PR DESCRIPTION
**Pull request by carols10cents**
There are some parts of the tests that should probably be part of the `arrow-flight` crate instead, but that can be done in a future PR. Everything here is in the integration tests.

### CODE REVIEW

File: rust/integration-testing/src/bin/flight-test-integration-client.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        .expect("No response received")
+        .expect("Invalid response received");
+
+    let body = String::from_utf8(r.body).unwrap();
+    assert_eq!(body, AUTH_USERNAME);
+
+    Ok(())
+}
+
+// TODO: should this be extended, abstracted, and moved out of test code and into production code?
```
Comment (nevi-me) [inline rust/integration-testing/src/bin/flight-test-integration-client.rs:187]: my intuition says no, as I'm assuming that there'd be different ways of implementing authentication.
I have however not used Flight, so if the basic auth scenario is always this, we could move this into `arrow-flight`.

File: rust/integration-testing/src/bin/flight-test-integration-server.rs [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+    type DoPutStream = TonicStream<Result<PutResult, Status>>;
+    type DoActionStream = TonicStream<Result<arrow_flight::Result, Status>>;
+    type ListActionsStream = TonicStream<Result<ActionType, Status>>;
+    type DoExchangeStream = TonicStream<Result<FlightData, Status>>;
+
+    async fn get_schema(
+        &self,
+        _request: Request<FlightDescriptor>,
+    ) -> Result<Response<SchemaResult>, Status> {
+        Err(Status::unimplemented("Not yet implemented"))
```
Comment (nevi-me) [inline rust/integration-testing/src/bin/flight-test-integration-server.rs:71]: Are these NYI service methods due to them not being included in the integration testing?

**Review by nevi-me — APPROVED** [review]
LGTM subject to CI, just comments that don't need any further action beyond response/confirmation

File: rust/integration-testing/src/bin/flight-test-integration-client.rs (hunk 1 above)
Comment (carols10cents) [reply rust/integration-testing/src/bin/flight-test-integration-client.rs:187]: The C++ Flight client provides [this `Authenticate` function](https://github.com/apache/arrow/blob/cf243d4bfc496dedb21a9c155eaba10664266923/cpp/src/arrow/flight/client.cc#L977-L994) and the ability to specify a [`ClientAuthHandler`](https://github.com/apache/arrow/blob/dbeab70863c4d0cd3da800f18f7e474da624bb5f/cpp/src/arrow/flight/client_auth.h#L45-L59), but I'm not sure if that's the abstraction desired in Rust 🤷‍♀️

File: rust/integration-testing/src/bin/flight-test-integration-server.rs (hunk 2 above)
Comment (carols10cents) [reply rust/integration-testing/src/bin/flight-test-integration-server.rs:71]: Yes, there's nothing I can see that calls these methods in the integration tests.

**Comment by carols10cents** [comment]
The archery unit tests [are failing with this error](https://github.com/apache/arrow/pull/8641/checks?check_run_id=1386664293#step:7:19):

```
E   ImportError: cannot import name 'HTTPHeaderDict'
```

[In the step before the unit tests are run, there's this](https://github.com/apache/arrow/pull/8641/checks?check_run_id=1386664293#step:6:106):

```
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.
Successfully installed MarkupSafe-1.1.1 archery attrs-20.3.0 certifi-2020.11.8 cffi-1.14.3 chardet-3.0.4 click-7.1.2 cryptography-3.2.1 defusedxml-0.6.0 deprecated-1.2.10 gitdb-4.0.5 gitpython-3.1.11 idna-2.10 importlib-metadata-2.0.0 iniconfig-1.1.1 jinja2-2.11.2 jira-2.0.0 numpy-1.18.5 oauthlib-3.1.0 packaging-20.4 pandas-0.25.3 pathlib2-2.3.5 pbr-5.5.1 pluggy-0.13.1 py-1.9.0 pycparser-2.20 pygithub-1.53 pyjwt-1.7.1 pyparsing-2.4.7 pytest-6.1.2 python-dateutil-2.8.1 python-dotenv-0.15.0 pytz-2020.4 requests-2.24.0 requests-oauthlib-1.3.0 requests-toolbelt-0.9.1 responses-0.12.0 ruamel.yaml-0.16.12 ruamel.yaml.clib-0.2.2 semver-2.13.0 six-1.15.0 smmap-3.0.4 toml-0.10.2 toolz-0.11.1 urllib3-1.26.0 wrapt-1.12.1 zipp-1.2.0

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

requests 2.24.0 requires urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1, but you'll have urllib3 1.26.0 which is incompatible.
```

I'm not very familiar with Python but given it's November this seems likely...? I'm not sure the correct way to fix this, and I don't see any open PRs about it from a quick skim?

The [other failing test job says it's out of disk space](https://github.com/apache/arrow/pull/8641/checks?check_run_id=1386664394#step:8:1499), is there anything I can do about that?

**Comment by carols10cents** [comment]
Ok, so, now the integration test job got cancelled after 360 min, and suspiciously it appears to be cancelled [during the Flight tests](https://github.com/apache/arrow/pull/8641/checks?check_run_id=1396352052#step:8:8481) after C++ -> C++ and C++ -> Java, right when I would expect C++ -> Rust to be, sooooo it's possible that it's my fault somehow. I will investigate more on Monday.

File: rust/integration-testing/src/bin/flight-test-integration-client.rs (hunk 1 above)
Comment (alamb) [reply rust/integration-testing/src/bin/flight-test-integration-client.rs:187]: I think we can always add it later to the core Rust library -- for now, having code in the example that does the work that people can use / follow seems pretty good to me

**Review by alamb — COMMENTED** [review]
This is really cool @carols10cents . Thank you!

One very minor suggestion is to update https://github.com/integer32llc/arrow/blob/flight-integration-tests/rust/integration-testing/README.md to include mention of the new binaries, flight-test-integration

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Once the integration tests are passing, I think this PR is good to merge

**Review by jorgecarleitao — APPROVED** [review]
Really impressive, @carols10cents . Thank you so much for this ❤️

**Comment by carols10cents** [comment]
I've decided to stop wasting CI time on this because it's not nearly ready :) Will reopen a new PR when it is.

**Comment by jorgecarleitao** [comment]
Just to say that I really appreciate that you are working on this. Thank you so much  ❤️

**Pull request closed without merging**